lib/cli-runner/bin.ts

```diff
@@ -0,0 +1,428 @@
+	.option('watch', {
+		alias: 'w',
+		type: 'boolean',
+		description: 'Watch files in the current working directory for changes',
+		group: fileHandling,
+	})
+	.option('watch-files', {
+		type: 'array',
+		description: 'List of paths or globs to watch',
+		group: fileHandling,
+	})
+	.option('watch-ignore', {
+		type: 'array',
+		description: 'List of paths or globs to exclude from watching',
+		group: fileHandling,
+	})
+	//#endregion
+	//#region Test Filters
+	.option('fgrep', {
+		type: 'string',
+		alias: 'f',
+		description: 'Only run tests containing this string',
+		group: testFilters,
+	})
+	.option('grep', {
+		type: 'string',
+		alias: 'g',
+		description: 'Only run tests matching this string or regexp',
+		group: testFilters,
+	})
+	.option('invert', {
+		alias: 'i',
+		type: 'boolean',
+		description: 'Inverts --grep and --fgrep matches',
+		group: testFilters,
+	})
+	.parseSync();
+
+// Avoid TS rewriting async import into a require:
+type ConfigOrArray = TestConfiguration | TestConfiguration[];
+
+const configFileRules: { [ext: string]: (path: string) => Promise<ConfigOrArray | Promise<ConfigOrArray>> } = {
+	json: (path: string) => fs.readFile(path, 'utf8').then(JSON.parse),
+	js: (path) => require(path),
+	mjs: (path) => import(path),
+};
+
+interface IConfigWithPath {
+	config: TestConfiguration;
+	path: string;
+}
+
+class CliExpectedError extends Error {}
+
+main();
+
+async function main() {
+	let code = 0;
+
+	try {
+		let configs =
+			args.config !== configFileDefault ? await tryLoadConfigFile(args.config) : await loadDefaultConfigFile();
+
+		if (args.label?.length) {
+			configs = args.label.map((label) => {
+				const found = configs.find((c, i) => (typeof label === 'string' ? c.config.label === label : i === label));
+				if (!found) {
+					throw new CliExpectedError(`Could not find a configuration with label "${label}"`);
+				}
+				return found;
+			});
+		}
+
+		if (args.watch) {
+			await watchConfigs(configs);
+		} else {
+			code = await runConfigs(configs);
+		}
+	} catch (e) {
+		code = 1;
+		if (e instanceof CliExpectedError) {
+			console.error(e.message);
+		} else {
+			console.error((e as Error).stack || e);
+		}
+	} finally {
+		process.exit(code);
+	}
+}
+
+const WATCH_RUN_DEBOUNCE = 500;
+
+async function watchConfigs(configs: readonly IConfigWithPath[]) {
+	let debounceRun: NodeJS.Timer;
+	let rerun = false;
+	let running = true;
+	const runOrDebounce = () => {
+		if (debounceRun) {
+			clearTimeout(debounceRun);
+		}
+
+		debounceRun = setTimeout(async () => {
+			running = true;
+			rerun = false;
+			try {
+				await runConfigs(configs);
+			} finally {
+				running = false;
+				if (rerun) {
+					runOrDebounce();
+				}
+			}
+		}, WATCH_RUN_DEBOUNCE);
+	};
+
+	const watcher = chokidar.watch(args.watchFiles?.length ? args.watchFiles.map(String) : process.cwd(), {
+		ignored: ['**/.vscode-test/**', '**/node_modules/**', ...(args.watchIgnore || []).map(String)],
+		ignoreInitial: true,
+	});
```

Comment:
You might not need chokidar. Node.js has a [builtin file watcher](https://nodejs.org/api/fs.html#fswatchfilename-options-listener). I haven’t used this myself, so I’m not sure if that’s sufficient.

Reply:
There's also [@parcel/watcher](https://www.npmjs.com/package/@parcel/watcher) that we use in VS Code. The choice of Chokidar here was pure laziness; neither fs.watcher nor Parcel's watcher support watching based on glob patterns like Chokidar does. (Chokidar is also what Mocha uses under the hood for its implementation of --watch)